admin/sql/updates/20211216-mbs-12140-12141.sql

```diff
@@ -30,16 +41,381 @@ BEGIN
   DELETE FROM label_tag WHERE tag = any(empty_tag_ids);
   DELETE FROM place_tag WHERE tag = any(empty_tag_ids);
   DELETE FROM recording_tag WHERE tag = any(empty_tag_ids);
-  DELETE FROM release_group_tag WHERE tag = any(empty_tag_ids);
   DELETE FROM release_tag WHERE tag = any(empty_tag_ids);
+  DELETE FROM release_group_tag WHERE tag = any(empty_tag_ids);
   DELETE FROM series_tag WHERE tag = any(empty_tag_ids);
   DELETE FROM work_tag WHERE tag = any(empty_tag_ids);
+
+  -- delete_unused_tag would normally kick in to delete these, but not if
+  -- they were completely unreferenced prior to running this script.
+  DELETE FROM tag WHERE id = any(empty_tag_ids);
+
+  -- Find tags with uncontrolled whitespace and clean them up.
+  --
+  -- We may find that for any unclean tag, an existing tag with the
+  -- "cleaned up" name already exists.  In that case, we update all
+  -- *_tag_raw and *_tag rows to use the existing clean tag, and delete
+  -- the unclean one.
+  FOR ufw_tag IN (
+    SELECT * FROM tag WHERE NOT controlled_for_whitespace(name)
+  ) LOOP
+    RAISE NOTICE 'Tag with uncontrolled whitespace found: id=%, name=%',
+      ufw_tag.id, to_json(ufw_tag.name);
+
+    SELECT t2.id
+      FROM tag t1
+      JOIN tag t2
+        ON (t1.id = ufw_tag.id
+            AND t2.id != ufw_tag.id
+            AND t2.name = regexp_replace(btrim(t1.name), E'\\s{2,}', ' ', 'g'))
+      INTO existing_cfw_tag_id;
+
+    IF existing_cfw_tag_id IS NULL THEN
+      UPDATE tag
+         SET name = regexp_replace(btrim(name), E'\\s{2,}', ' ', 'g')
+       WHERE id = ufw_tag.id;
+    ELSE
+      RAISE NOTICE 'Conflicting tag with controlled whitespace found: id=%',
+        existing_cfw_tag_id;
+
+      OPEN tag_cursor NO SCROLL FOR SELECT * FROM area_tag_raw WHERE tag = ufw_tag.id FOR UPDATE;
```

Comment:
I tried to find a way to templatize this block and simply loop over the entity type names, but in the end it was easier to just copy and paste it for each type.

 * You can't use `WHERE CURRENT OF` inside `EXECUTE`.
 * The `tag_cursor.ctid` doesn't match the underlying row's `ctid`, so can't use that either.
 * We can finally use the primary keys to identify the current row.  You can't cast a record type to a table type directly, so something like `($1).%I` where `$1` is the current cursor record and `%I` is e.g. "release_group" doesn't work.  There is a workaround where you can cast to `::TEXT` first.  But the primary keys of the *_tag and *_tag_raw tables still differ, so ultimately the templatized code would be hard to read.